In [18]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import os

In [19]:
os.getcwd()

'C:\\Users\\90536'

In [20]:
os.listdir('C:\\Users\\90536\\Downloads\\train.csv')

['train.csv']

In [21]:

data = pd.read_csv(r'C:\Users\90536\Downloads\train.csv\train.csv')
data.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   
2                      0.0                      0.0  ...   
3                      0.0                      0.0  ...   
4                      0.0                      0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 371 columns]

In [22]:
X = data.drop(['TARGET'] , axis = 1)
Y = data["TARGET"]

In [23]:
# Train Test Split
X_train, X_test, y_train, y_test = train_test_split(X,Y,test_size=0.2,random_state=0)

In [24]:
X_train.shape, X_test.shape , y_train.shape , y_test.shape

((60816, 370), (15204, 370), (60816,), (15204,))

In [25]:
data.shape

(76020, 371)

In [26]:
# Original training and test data to compare with feature selected dataset
X_train_original = X_train.copy()
X_test_original = X_test.copy()

# Remove Constant Features from Dataset

In [27]:
#As they(Features) are CONSTANT thus play no role in CLASSIFYING the Output.Thus Remove Constant Features

constant_features = [features for features in X_train.columns if X_train[features].std() == 0]

In [28]:
#DROPPING CONSTANT FEATURES

#X_train = X_train.drop(labels = constant_features , axis = 1 )
#X_test = X_test.drop(labels = constant_features , axis = 1 )

X_train.drop(labels=constant_features, axis=1, inplace=True)
X_test.drop(labels=constant_features, axis=1, inplace=True)
 

C:\Users\90536\anaconda3\lib\site-packages\pandas\core\frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [29]:
X_train.shape, X_test.shape

((60816, 332), (15204, 332))

# Remove Quasi-Constant Features

As Quasi Constant Features are HAVE 99% Similiar Value for ALL DIFFERENT TYPES of CLASSES/Outputs so these Features can't DECIDE/classify the Output. Thus Remove them

In [30]:
from sklearn.feature_selection import VarianceThreshold

FS = VarianceThreshold(threshold=0.01)
FS.fit(X_train)

VarianceThreshold(threshold=0.01)

In [31]:
from sklearn.feature_selection import VarianceThreshold

FS = VarianceThreshold(threshold=0.01)
FS.fit(X_train)

VarianceThreshold(threshold=0.01)

In [32]:
print("Number of NON-Quasi Features are: {} ".format(sum(FS.get_support())))

Number of NON-Quasi Features are: 267 


In [33]:
FS.get_support()[:20]

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

In [34]:
#Selecting Only NON-QUASI FEATURES from Train Data
X_train = X_train.iloc[: , FS.get_support()]
X_train.shape

(60816, 267)

In [35]:
#Selecting Only NON-QUASI FEATURES from Test Data

X_test = X_test.iloc[: , FS.get_support()]
X_test.shape


(15204, 267)

# Remove Duplicate Features

In [36]:
duplicated_features = []
for i in range(0, len(X_train.columns)):
    
 
    col_1 = X_train.columns[i]
 
    for col_2 in X_train.columns[i + 1:]:
        if X_train[col_1].equals(X_train[col_2]):
            duplicated_features.append(col_2)
            
len(duplicated_features)

16

In [37]:
# Remove duplicated features
X_train.drop(labels=duplicated_features, axis=1, inplace=True)
X_test.drop(labels=duplicated_features, axis=1, inplace=True)
 

In [38]:
X_train.shape, X_test.shape

((60816, 251), (15204, 251))

In [39]:
X_train_basic_filter = X_train.copy()
X_test_basic_filter = X_test.copy()

# Remove Features Having High Correlation

In [40]:
def correlation(dataset, threshold):
    col_corr = set()  # Set of all the names of correlated columns
    corr_matrix = dataset.corr()
    for i in range(len(corr_matrix.columns)):
        for j in range(i):
            if abs(corr_matrix.iloc[i, j]) > threshold: # we are interested in absolute coeff value
                colname = corr_matrix.columns[i]  # getting the name of column
                col_corr.add(colname)
    return col_corr
 
corr_features = correlation(X_train, 0.8)
print('Number of correlated features  ', len(set(corr_features)) )

Number of correlated features   128


In [41]:
X_train.drop(labels=corr_features, axis=1, inplace=True)
X_test.drop(labels=corr_features, axis=1, inplace=True)
 
X_train.shape, X_test.shape

((60816, 123), (15204, 123))

# Applying Machine Learning Models:

In [42]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score


# LOGISTIC Regression Algorithm

In [43]:
# Function to build logistic regression and compare performance in train and test set
def run_logistic(X_train, X_test, y_train, y_test):
    # function to train and test the performance of logistic regression
    logit = LogisticRegression(random_state=44)
    logit.fit(X_train, y_train)
    print('Train set')
    pred = logit.predict_proba(X_train)
    print('Logistic Regression roc-auc: {}'.format(roc_auc_score(y_train, pred[:,1])))
    print('Test set')
    pred = logit.predict_proba(X_test)
    print('Logistic Regression roc-auc: {}'.format(roc_auc_score(y_test, pred[:,1])))


In [44]:
# Performance of Logistic Regression Classifier on Original Dataset

scaler = StandardScaler().fit(X_train_original.drop(labels=['ID'], axis=1))
 
run_logistic(scaler.transform(X_train_original.drop(labels=['ID'], axis=1)),
             scaler.transform(X_test_original.drop(labels=['ID'], axis=1)),
                  y_train, y_test)

C:\Users\90536\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


Train set
Logistic Regression roc-auc: 0.8019421187812357
Test set
Logistic Regression roc-auc: 0.8101611062877137


In [45]:
# Performance of Logistic Regression Classifier on Dataset without Constants, Quasi-Constants and Duplicates
scaler = StandardScaler().fit(X_train_basic_filter.drop(labels=['ID'], axis=1))
 
run_logistic(scaler.transform(X_train_basic_filter.drop(labels=['ID'], axis=1)),
             scaler.transform(X_test_basic_filter.drop(labels=['ID'], axis=1)),
                  y_train, y_test)

Train set
Logistic Regression roc-auc: 0.8013181878031596
Test set
Logistic Regression roc-auc: 0.8090590622016199


C:\Users\90536\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [46]:
# Performance of Logistic Regression Classifier on Dataset without Correaltion between Features
scaler = StandardScaler().fit(X_train.drop(labels=['ID'], axis=1))
 
run_logistic(scaler.transform(X_train.drop(labels=['ID'], axis=1)),
             scaler.transform(X_test.drop(labels=['ID'], axis=1)),
                  y_train, y_test)

Train set
Logistic Regression roc-auc: 0.7934889045445643
Test set
Logistic Regression roc-auc: 0.8041125838711841


C:\Users\90536\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### Model faced CONVERGENCE PROBLEM as THERE WERE 371 , 251 and 123 number of Features (BEFORE FEATURE SELECTION and preprocessing we had 371 and after we got 123 Features). As FEATURES were a lot so we faced this Convergence Problem

# SUPPORT VECTOR MACHINE SVM Algorithm

In [47]:
#Import svm model
from sklearn import svm

from sklearn import metrics

def run_SVM(X_train, X_test, y_train, y_test):
    # function to train and test the performance of SVM
    
    #Create a SVM Classifier
    clf = svm.SVC(kernel='linear', probability=True) # Linear Kernel

    clf.fit(X_train, y_train)
    print('Train set')
    pred = clf.predict_proba(X_train)
    print('SVM roc-auc: {}'.format(roc_auc_score(y_train, pred[:,1])))
    #print('Test set')
    #pred = clf.predict_proba(X_test)
    #print('SVM roc-auc: {}'.format(roc_auc_score(y_test, pred[:,1])))
    from sklearn import metrics

    print("Accuracy:",metrics.accuracy_score(y_test, pred[:,1]))

In [48]:
from sklearn.decomposition import PCA

pca = PCA(n_components = None)

X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

variance = pca.explained_variance_ratio_
variance

array([6.26024084e-01, 1.64582700e-01, 6.20140896e-02, 5.79788739e-02,
       3.37266829e-02, 2.93163913e-02, 1.15111181e-02, 4.97237184e-03,
       4.45820422e-03, 3.77269944e-03, 1.64272458e-03, 4.28711382e-08,
       7.30819481e-09, 2.60244126e-09, 2.21670839e-09, 1.98067634e-09,
       8.67661752e-10, 7.94299586e-10, 4.18902780e-10, 3.17442055e-10,
       1.99691543e-10, 1.12270142e-10, 9.00306100e-11, 8.56134469e-11,
       8.11586760e-11, 8.11269521e-11, 7.37103734e-11, 6.43706328e-11,
       4.97805387e-11, 4.19396128e-11, 3.08481302e-11, 2.72862146e-11,
       1.50344706e-11, 1.22125491e-11, 1.05433778e-11, 1.02526731e-11,
       5.18255112e-12, 5.06486133e-12, 3.67630579e-12, 2.87184576e-12,
       2.41014808e-12, 2.09319356e-12, 1.50627547e-12, 1.20014634e-12,
       7.34651930e-13, 5.91943946e-13, 4.03005004e-13, 3.54423203e-13,
       3.52630064e-13, 3.17785814e-13, 2.53718824e-13, 2.47120601e-13,
       2.31058285e-13, 2.06603099e-13, 2.03434286e-13, 1.18563420e-13,
      

In [49]:
pca = PCA(n_components = 2)

X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

X_train = pd.DataFrame(X_train)
X_test = pd.DataFrame(X_test)

In [55]:
# Performance of SVM Classifier on Dataset without Correaltion between Features
#scaler = StandardScaler().fit(X_train)
 
#run_SVM(scaler.transform(X_train),
#             scaler.transform(X_test),
#                  y_train, y_test)

# RANDOM FOREST CLASSIFIER

In [51]:
# Function to build Random Forest Classifier and compare performance in train and test set
def run_randomForests(X_train, X_test, y_train, y_test):
    rf = RandomForestClassifier(n_estimators=200, random_state=39, max_depth=4)
    rf.fit(X_train, y_train)
    print('Train set')
    pred = rf.predict_proba(X_train)
    print('Random Forests roc-auc: {}'.format(roc_auc_score(y_train, pred[:,1])))
    print('Test set')
    pred = rf.predict_proba(X_test)
    print('Random Forests roc-auc: {}'.format(roc_auc_score(y_test, pred[:,1])))

In [52]:
# Performance of RFC on Original Dataset with Constants, Duplicates and Correlated Features
run_randomForests(X_train_original.drop(labels=['ID'], axis=1),
                  X_test_original.drop(labels=['ID'], axis=1),
                  y_train, y_test)

Train set
Random Forests roc-auc: 0.7996929040052009
Test set
Random Forests roc-auc: 0.8063437901223074


In [53]:
# Performance of RFC on Dataset without any Constants or Duplicates
run_randomForests(X_train_basic_filter.drop(labels=['ID'], axis=1),
                  X_test_basic_filter.drop(labels=['ID'], axis=1),
                  y_train, y_test)

Train set
Random Forests roc-auc: 0.8004242110792267
Test set
Random Forests roc-auc: 0.8058406171446881
